glean_parser/rust.py

```diff
@@ -0,0 +1,199 @@
+                yield "vec!["
+                first = True
+                for subvalue in sorted(list(value)):
+                    if not first:
+                        yield ", "
+                    yield from self.iterencode(subvalue)
+                    first = False
+                yield "]"
+            elif isinstance(value, list):
+                yield "vec!["
+                first = True
+                for subvalue in list(value):
+                    if not first:
+                        yield ", "
+                    yield from self.iterencode(subvalue)
+                    first = False
+                yield "]"
+            elif value is None:
+                yield "None"
+            elif isinstance(value, str):
+                yield '"' + value + '".into()'
+            else:
+                yield from super().iterencode(value)
+
+    return "".join(RustEncoder().iterencode(value))
+
+
+def ctor(obj):
+    """
+    Returns the scope and name of the constructor to use for a metric object.
+    Necessary because LabeledMetric<T> is constructed using LabeledMetric::new
+    not LabeledMetric<T>::new
+    """
+    if getattr(obj, "labeled", False):
+        return "LabeledMetric::new"
+    return class_name(obj.type) + "::new"
+
+
+def type_name(obj):
+    """
+    Returns the Rust type to use for a given metric or ping object.
+    """
+
+    if getattr(obj, "labeled", False):
+        return "LabeledMetric<Labeled{}>".format(class_name(obj.type))
+    generate_enums = getattr(obj, "_generate_enums", [])  # Extra Keys? Reasons?
+    if len(generate_enums):
+        for name, suffix in generate_enums:
+            if not len(getattr(obj, name)) and suffix == "Keys":
+                return class_name(obj.type) + "<NoExtraKeys>"
+            else:
+                # we always use the `extra` suffix,
+                # because we only expose the new event API
+                suffix = "Extra"
+                return "{}<{}>".format(
+                    class_name(obj.type), util.Camelize(obj.name) + suffix
+                )
+    return class_name(obj.type)
+
+
+def extra_type_name(typ: str) -> str:
+    """
+    Returns the corresponding Rust type for event's extra key types.
+    """
+
+    if typ == "boolean":
+        return "bool"
+    elif typ == "string":
+        return "String"
+    elif typ == "quantity":
+        return "u32"
+    else:
+        return "UNSUPPORTED"
+
+
+def class_name(obj_type):
+    """
+    Returns the Rust class name for a given metric or ping type.
+    """
+    if obj_type == "ping":
+        return "Ping"
+    if obj_type.startswith("labeled_"):
+        obj_type = obj_type[8:]
+    return util.Camelize(obj_type) + "Metric"
+
+
+def extra_keys(allowed_extra_keys):
+    """
+    Returns the &'static [&'static str] ALLOWED_EXTRA_KEYS for impl ExtraKeys
+    """
+    return "&[" + ", ".join(map(lambda key: '"' + key + '"', allowed_extra_keys)) + "]"
+
+
+class Category:
```

Comment:
Maybe slightly better to use a constructor (`__init__`) here rather than setting the methods directly.  (Practically, since this is kind of a private class it doesn't matter much, but it would if this were a public API).  See https://github.com/mozilla/glean_parser/blob/main/glean_parser/metrics.py#L36 for an example.